Bluetooth: Host: Fixes for "DB Out of Sync"

hermabe opened the pull request:
This patch makes sure the host only sends the "Database Out Of Sync" error once per channel, as per the spec:
> BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 3, Part G, 2.5.2.1 Robust Caching
>
> The ATT_ERROR_RSP PDU is sent only once per bearer after the client
becomes change-unaware, unless the client disconnects or the database
changes again before the client becomes change-aware in which case the
ATT_ERROR_RSP PDU shall be sent again.

Also adds a test that verifies that the error is not sent on repeated reads.

hermabe committed:
Bluetooth: Host: Update GATT caching test to encrypt the link

EATT channels shall be encrypted. Without encrypting the link, all
requests will be sent on the fixed ATT channel.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Tests: Wait for sync from other devices only

The sync mechanism using Babblesim backchannels picked up self-sent
messages. Use the device number as data to disambiguate messages from
other devices.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Add more tests for GATT client becoming change-aware

For clients with both a single and multiple ATT channels:
 - Read a characteristic before reading the DB hash and then retrying
 - Read the DB hash and then do the reads
 - Retry the reads without reading the DB hash

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/common.c:
```diff
@@ -39,16 +40,25 @@ void backchannel_init(void)
+	while (true) {
+		if (bs_bc_is_msg_received(CHANNEL_ID) > 0) {
+			bs_bc_receive_msg(CHANNEL_ID, sync_msg, ARRAY_SIZE(sync_msg));
+			if (sync_msg[0] != get_device_nbr()) {
```
Just out of curiosity since I can't find it documented anywhere: what is `nbr` in this context? :D

hermabe replied:
`get_device_nbr()` returns the "Overall device number in the simulation", which is set by the "-d" or "-gd" option when running the test executable. In typical C fashion, all vowels are removed.

hermabe replied:
Had to search through to `babblesim/components/libUtilv1/src/bs_cmd_line_typical.h` to find the definition.

Thalley replied:
Ah, yes
```
    { false, true  , false, "d",       "device_number",        'u', (void*)&args->device_nbr,NULL,         "Device number (for this phy)"}
```

I think this is the first time I've encountered the `number`->`nbr` abbreviation :D

sjanc commented:
#AutoPTS run zephyr GATT/SR/GAS/BV-06-C

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/44648#issuecomment-1094912264 after 13:06:45.

codecoup-tester commented:
AutoPTS Bot results:
No regressions found.
No progresses found.

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/44648#issuecomment-1094912264 after 13:25:52.

codecoup-tester commented:
AutoPTS Bot results:
Failed tests:
GATT    GATT/SR/GAS/BV-06-C             FAIL
No regressions found.
No progresses found.

hermabe committed:
Bluetooth: Host: Send the "Out of Sync" error only once on each channel

The "Database Out Of Sync" error response shall only be sent once on
each channel after a client becomes change-unaware.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Add CF_DB_HASH_READ flag

This fixes the case where the client reads the DB hash without reading
another attribute first after becoming change-unaware.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented:
Made some changes to incorporate a fix for https://github.com/zephyrproject-rtos/zephyr/issues/44734. Added more testing to ensure the fix works in more cases.

hermabe commented:
#AutoPTS run zephyr GATT/SR/GAS/BV-06-C

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3478,3 +3499,31 @@ bool bt_att_fixed_chan_only(struct bt_conn *conn)
+
+bool bt_att_out_of_sync_sent_on_fixed(struct bt_conn *conn)
+{
+	struct bt_l2cap_chan *chan;
+	struct bt_att_chan *att_chan;
+
+	chan = bt_l2cap_le_lookup_rx_cid(conn, BT_L2CAP_CID_ATT);
+	if (!chan) {
+		return false;
+	}
+
+	att_chan = ATT_CHAN(chan);
```
```suggestion
	struct bt_l2cap_chan *l2cap_chan;
	struct bt_att_chan *att_chan;

	l2cap_chan; = bt_l2cap_le_lookup_rx_cid(conn, BT_L2CAP_CID_ATT);
	if (!l2cap_chan;) {
		return false;
	}

	att_chan = ATT_CHAN(l2cap_chan;);
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5259,6 +5262,19 @@ bool bt_gatt_change_aware(struct bt_conn *conn, bool req)
+	 * the Database Hash characteristic and then the server receives another
+	 * ATT request from the client.
+	 */
+	if (atomic_test_bit(cfg->flags, CF_DB_HASH_READ)) {
+		atomic_clear_bit(cfg->flags, CF_DB_HASH_READ);
```
```suggestion
	if (atomic_test_and_clear_bit(cfg->flags, CF_DB_HASH_READ)) {
```

Thalley reviewed (dismissed):
Two minor suggestions, otherwise LGTM

sjanc commented:
#AutoPTS run zephyr GATT/SR/GAS/BV-06-C

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
No regressions found.
No progresses found.

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/44648#issuecomment-1096787930 after 16:29:15.

carlescufi commented:
@sjanc could you approve?
